File: src/lib/matching.ts
import type {
  Bar,
  Coords,
  ManhattanNeighborhood,
  VibeProfile,
  VibeTag,
} from '@/types';
import { haversineMiles } from '@/lib/distance';
import {
  EMPTY_TASTE,
  learnedTasteScore,
  type LearnedTaste,
} from '@/lib/tasteAffinity';
import { daysAgo } from '@/lib/freshness';
import { nycHour } from '@/lib/nightKey';
import {
  LATE_CLUB_BOOST,
  LATE_NIGHT_END_HOUR,
  LATE_NIGHT_START_HOUR,
  LATE_RESTAURANT_PENALTY,
  EXPLORATION_MIN_RESULTS,
  LAST_VERIFIED_HARD_FILTER_DAYS,
  RADIUS_CAB,
  RADIUS_WALK,
  MAX_RESULTS,
} from '@/lib/constants';

export function jaccard(a: VibeTag[], b: VibeTag[]): number {
  if (a.length === 0 && b.length === 0) return 0;
  const setA = new Set(a);
  const setB = new Set(b);
  let intersection = 0;
  for (const tag of setA) {
    if (setB.has(tag)) intersection += 1;
  }
  const union = setA.size + setB.size - intersection;
  if (union === 0) return 0;
  return intersection / union;
}

/**
 * The vibes the user EXPLICITLY selected, deduplicated — or null when no
 * selection is active (V8-R-NXT-009 / D-C-41).
 *
 * The distinction is the whole point: a SAVED QUIZ PROFILE is a cold-start
 * prior, not a choice the user just made, so it never reaches here. Only an
 * APPLIED Tweak-the-vibe pick sets `isExplicitVibe`, and an applied EMPTY pick
 * is a CLEAR — it returns null, which restores ungated ranking and removes the
 * badge exactly as if the surface had never been opened.
 *
 * Deduplication happens once, here, so N is the number of DISTINCT vibes for
 * both the badge denominator and the eligibility threshold.
 */
export function selectedVibes(profile: VibeProfile): VibeTag[] | null {
  if (profile.isExplicitVibe !== true) return null;
  const unique = [...new Set(profile.tags)];
  return unique.length > 0 ? unique : null;
}

/** How many of the SELECTED vibes a bar carries. Both sides deduplicated. */
export function vibeMatchCount(selected: VibeTag[], barTags: VibeTag[]): number {
  const bar = new Set(barTags);
  let num = 0;
  for (const tag of new Set(selected)) {
    if (bar.has(tag)) num += 1;
  }
  return num;
}

/**
 * ELIGIBILITY (V8-R-NXT-009 / D-C-41): with N distinct selected vibes a bar
 * must carry at least max(1, N - 1) of them. So 1/2, 3/4 and 5/6 are admitted
 * — one miss is forgiven — while 0/1, 0/2, 2/4 and 4/6 are not.
 *
 * With NO selection this gates nothing: quiz cold-start and learned taste are
 * untouched, and no bar is rejected for tag mismatch.
 */
export function isVibeEligible(selected: VibeTag[], barTags: VibeTag[]): boolean {
  const den = new Set(selected).size;
  if (den === 0) return true;
  return vibeMatchCount(selected, barTags) >= Math.max(1, den - 1);
}

/**
 * The match badge, or NULL when there is nothing honest to show.
 *
 * The denominator is N — the number of vibes the user actually picked — not
 * `min(|user|, |bar|)`, which quietly shrank the denominator to flatter a bar
 * with few tags. Returning null (rather than 0/1) for an empty selection is
 * what makes "no explicit selection, no badge" structural: the card cannot
 * render a badge it was never given.
 */
export function vibeMatchBadge(
  selected: VibeTag[],
  barTags: VibeTag[],
): { num: number; den: number } | null {
  const den = new Set(selected).size;
  if (den === 0) return null;
  return { num: vibeMatchCount(selected, barTags), den };
}

export type MatchesArgs = {
  profile: VibeProfile;
  coords: Coords | null;
  preferredNeighborhoods: ManhattanNeighborhood[];
  /** Exclusive lower edge for a distance band; null keeps nearby bars. */
  minMilesExclusive?: number | null;
  maxMiles: number | null;
  /** Broader travel searches rank the admitted pool together; exact miles still break ties. */
  distanceBands?: boolean;
  bars: Bar[];
  excludeIds?: string[];
  maxResults?: number;
  now?: Date;
  /**
   * Live-surface clock for the LATE-NIGHT bias (operator 2026-07-27):
   * between LATE_NIGHT_START_HOUR and LATE_NIGHT_END_HOUR, club/dance
   * venues get a small additive boost and restaurant-bars a small
   * penalty — nightlife first when it's nightlife o'clock. Omitted on
   * quiz/planning surfaces (they browse, not "right now").
   */
  biasNow?: Date;
  /**
   * Learned taste from numeric scores (V8 P1). Omitted = no rating history,
   * so c = 0 and the quiz prior alone orders the page — the cold-start case.
   */
  taste?: LearnedTaste;
};

/**
 * 10pm–3:59am in NEW YORK — when the night bias applies.
 *
 * getHours() answered in the device's zone, so a user in Los Angeles got the
 * club boost and the restaurant penalty three hours off (round-3 panel, Codex).
 * This is an NYC-only matcher; the hour comes from the same module as the
 * rollover, so there is one clock here and nowhere else.
 */
export function isLateNight(now: Date): boolean {
  const h = nycHour(now);
  if (Number.isNaN(h)) return false; // broken clock: no bias rather than a wrong one
  return h >= LATE_NIGHT_START_HOUR || h < LATE_NIGHT_END_HOUR;
}

/**
 * Additive late-night nudge at tie-breaker scale: genuine nightlife
 * (club/dance) up, restaurant-bars down — unless the venue is BOTH (a
 * restaurant that becomes a club keeps its night credibility).
 */
export function lateNightAdjustment(bar: Pick<Bar, 'tags'>): number {
  const isClub = bar.tags.includes('club') || bar.tags.includes('dance');
  if (isClub) return LATE_CLUB_BOOST;
  if (bar.tags.includes('restaurant-bar')) return -LATE_RESTAURANT_PENALTY;
  return 0;
}

/**
 * Explicit-intent weighting — APPROVED product assumption (operator,
 * 2026-08-19). Do not re-tune silently; a change here is a product decision.
 *
 * An APPLIED Tweak-the-vibe pick is an instruction the user just gave, not a
 * prior to be shrunk away. Routed through the quiz term it was weighted
 * (1 - c), so at 200 ratings (c = 0.952) an explicit pick carried 4.8% of the
 * ranking — a matching cocktail bar lost to a nonmatching pub the user simply
 * had more history with. On the explicit path the split is FIXED at 80/20 and
 * independent of c.
 *
 * This is a SEPARATE path, not a re-tuning of the (1 - c) prior: with no
 * tweak applied, rankScore and the band walk below behave exactly as before.
 */
const EXPLICIT_VIBE_WEIGHT = 0.8;
const EXPLICIT_TASTE_WEIGHT = 1 - EXPLICIT_VIBE_WEIGHT;

/**
 * A bar's WITHIN-BAND ordering value — the cascade's step 3.
 *
 *   c·learnedTaste + (1 - c)·quizPrior  (+ the late-night nudge)
 *
 * c = N/(N+10) shrinks the quiz prior away as real rating history arrives, so
 * quiz tags are a cold-start prior and never a permanent weighted term.
 * Distance is deliberately ABSENT: it selects the band and breaks ties, it is
 * not a ranking term. Exported as the offline-eval oracle (it replaces the
 * removed scoreBar in that role) — matches() is its only production caller.
 *
 * Used when NO vibe tweak is active; an applied tweak ranks by
 * explicitVibeScore instead.
 */
export function rankScore(
  bar: Bar,
  quizTags: VibeTag[],
  taste: LearnedTaste,
  late = false,
): number {
  const c = taste.confidence;
  return (
    c * learnedTasteScore(bar, taste) +
    (1 - c) * jaccard(quizTags, bar.tags) +
    (late ? lateNightAdjustment(bar) : 0)
  );
}

/**
 * rankScore's counterpart for an ACTIVE vibe tweak:
 *
 *   0.8·vibeMatch + 0.2·learnedTaste  (+ the same late-night nudge)
 *
 * No confidence term — that is the whole point. Learned taste still shapes
 * the order (it breaks ties among equally matching bars, and pushes a bar the
 * user has scored badly down), but it can no longer outvote what the user
 * just asked for. Exported alongside rankScore as the offline-eval oracle for
 * the explicit path.
 */
export function explicitVibeScore(
  bar: Bar,
  vibeTags: VibeTag[],
  taste: LearnedTaste,
  late = false,
): number {
  return (
    EXPLICIT_VIBE_WEIGHT * jaccard(vibeTags, bar.tags) +
    EXPLICIT_TASTE_WEIGHT * learnedTasteScore(bar, taste) +
    (late ? lateNightAdjustment(bar) : 0)
  );
}

export function matches(args: MatchesArgs): Bar[] {
  const {
    profile,
    coords,
    preferredNeighborhoods,
    minMilesExclusive = null,
    maxMiles,
    distanceBands = true,
    bars,
    excludeIds,
    maxResults,
    now,
    biasNow,
    taste = EMPTY_TASTE,
  } = args;

  const exclude = new Set(excludeIds ?? []);
  let pool = bars
    .filter((b) => !exclude.has(b.id))
    // Never SUGGEST a dead bar: Places refresh (2026-07-23) found ~31
    // catalog bars permanently closed. Badge-only handling isn't enough —
    // the matcher must hard-filter them.
    .filter((b) => b.businessStatus !== 'CLOSED_PERMANENTLY')
    .filter((b) => daysAgo(b.lastVerified, now) <= LAST_VERIFIED_HARD_FILTER_DAYS);

  if (preferredNeighborhoods.length > 0) {
    const allowed = new Set(preferredNeighborhoods);
    pool = pool.filter((b) => allowed.has(b.neighborhood));
  }

  // The explicit-intent path (EXPLICIT_VIBE_WEIGHT). Active only for an
  // APPLIED tweak carrying at least one tag — an empty pick is a CLEARED
  // tweak, which must rank exactly like no tweak at all. Resolved BEFORE the
  // distance filter because an applied pick changes what that filter admits.
  const selected = selectedVibes(profile);

  // ELIGIBILITY, not ordering (V8-R-NXT-009 / D-C-41). This runs BEFORE the
  // distance filter, before banding, before the cap — and before ResultsView's
  // route-candidate truncation, which only ever sees what this returns. There
  // is deliberately no second pass that lets a rejected bar back in: the fill
  // order, the band expansion, "run it again" and the route supplement all
  // draw from this pool, so a bar the user's picks rejected cannot reappear as
  // padding. Fewer results — or none — is the honest answer.
  if (selected) {
    pool = pool.filter((b) => isVibeEligible(selected, b.tags));
  }

  // Explicit vibes never widen a caller's geographic bounds.
  if (coords && (minMilesExclusive !== null || maxMiles !== null)) {
    pool = pool.filter((b) => {
      const miles = haversineMiles(coords, b);
      return (minMilesExclusive === null || miles > minMilesExclusive) &&
        (maxMiles === null || miles <= maxMiles);
    });
  }

  const cap = maxResults ?? MAX_RESULTS;

  // ---- V8 P1 cascade (docs/V8-PRD-2026-08-13.md) --------------------------
  // Quiz tags are a COLD-START PRIOR, never an admission gate. The old
  // adaptive-Jaccard filter (JACCARD_START relaxing to JACCARD_FLOOR) is gone:
  // no bar is rejected for tag mismatch any more. Ordering carries taste, and
  // the quiz prior fades as c grows with rating history.
  const late = biasNow !== undefined && isLateNight(biasNow);

  const rankOf = selected
    ? (bar: Bar): number => explicitVibeScore(bar, selected, taste, late)
    : (bar: Bar): number => rankScore(bar, profile.tags, taste, late);

  // Step 2 — fill from the CLOSEST band first, expanding only when the closer
  // band cannot fill the page. Bands reuse the existing distance-chip
  // constants; inventing new thresholds here would be a new product
  // assumption. With no coords there is nothing to band on, so the whole pool
  // is one band (the pre-GPS behavior).
  const milesOf = coords
    ? (bar: Bar) => haversineMiles(coords, bar)
    : () => 0;
  const bands: Bar[][] = coords && distanceBands ? [[], [], []] : [pool];
  if (coords && distanceBands) {
    for (const bar of pool) {
      const mi = milesOf(bar);
      bands[mi <= RADIUS_WALK ? 0 : mi <= RADIUS_CAB ? 1 : 2].push(bar);
    }
  }

  // Step 2b USED to re-order each band's matching bars ahead of its
  // nonmatching ones so an active tweak expanded before it fell back. Under
  // D-C-41 there is nothing left to fall back TO — every bar that survived the
  // eligibility filter matches — so the bands are walked as they are.

  // Steps 3 + 4 — within a band, learned taste orders; EXACT MILES are only
  // the final tie-breaker, never a ranking term of their own.
  const ranked: { bar: Bar; score: number; miles: number }[] = [];
  for (const band of bands) {
    if (ranked.length >= cap) break;
    ranked.push(
      ...band
        .map((bar) => ({ bar, score: rankOf(bar), miles: milesOf(bar) }))
        .sort((a, b) => b.score - a.score || a.miles - b.miles),
    );
  }

  const top = ranked.slice(0, cap).map((r) => r.bar);

  // The B7b exploration slot used to overwrite the last result on 10+ slot
  // surfaces with a deterministic long-tail pick. REMOVED 2026-08-19 by
  // operator direction: V8 ranking correctness wins over the old exploration
  // behavior, and every returned result must be ordered by
  // distance band -> learned taste -> exact miles. The overwrite bypassed
  // steps 3 and 4 for the Map's tenth result. If exploration is wanted again
  // it returns as its own goal, designed to fit the cascade rather than to
  // punch a hole in it. The deleted implementation (explorationSeed, the
  // deterministic night-keyed pick and its evals) is recoverable from git at
  // e4299f4 — do not rebuild it from scratch.

  return top;
}

File: apple/Packages/NextBarCore/Sources/NextBarCore/Matching.swift
import Foundation

/// Ports `src/lib/matching.ts` — the ranker.

public func jaccard(_ a: [VibeTag], _ b: [VibeTag]) -> Double {
    if a.isEmpty && b.isEmpty { return 0 }
    let setA = Set(a)
    let setB = Set(b)
    let intersection = setA.intersection(setB).count
    let union = setA.count + setB.count - intersection
    if union == 0 { return 0 }
    return Double(intersection) / Double(union)
}

/// The vibes the user EXPLICITLY selected, deduplicated — or nil when no
/// selection is active (V8-R-NXT-009 / D-C-41). A saved quiz profile is a
/// cold-start prior, not a choice the user just made, so it never reaches
/// here. Only an APPLIED Tweak-the-vibe pick sets `isExplicitVibe`, and an
/// applied EMPTY pick is a CLEAR — it returns nil.
public func selectedVibes(_ profile: VibeProfile) -> [VibeTag]? {
    guard profile.isExplicitVibe == true else { return nil }
    var seen = Set<VibeTag>()
    var unique: [VibeTag] = []
    for tag in profile.tags where !seen.contains(tag) {
        seen.insert(tag)
        unique.append(tag)
    }
    return unique.isEmpty ? nil : unique
}

/// How many of the SELECTED vibes a bar carries. Both sides deduplicated.
public func vibeMatchCount(_ selected: [VibeTag], _ barTags: [VibeTag]) -> Int {
    let bar = Set(barTags)
    var num = 0
    for tag in Set(selected) where bar.contains(tag) { num += 1 }
    return num
}

/// ELIGIBILITY (V8-R-NXT-009 / D-C-41): with N distinct selected vibes a bar
/// must carry at least max(1, N - 1) of them — one miss is forgiven. With NO
/// selection this gates nothing.
public func isVibeEligible(_ selected: [VibeTag], _ barTags: [VibeTag]) -> Bool {
    let den = Set(selected).count
    if den == 0 { return true }
    return vibeMatchCount(selected, barTags) >= max(1, den - 1)
}

/// The match badge, or nil when there is nothing honest to show. The
/// denominator is N — the number of vibes the user actually picked — never
/// `min(|user|, |bar|)`.
public struct VibeMatchBadge: Equatable, Sendable {
    public let num: Int
    public let den: Int
}

public func vibeMatchBadge(_ selected: [VibeTag], _ barTags: [VibeTag]) -> VibeMatchBadge? {
    let den = Set(selected).count
    if den == 0 { return nil }
    return VibeMatchBadge(num: vibeMatchCount(selected, barTags), den: den)
}

public struct MatchesArgs {
    public var profile: VibeProfile
    public var coords: Coords?
    public var preferredNeighborhoods: [Neighborhood]
    /// Exclusive lower edge for a distance band; nil keeps nearby bars.
    public var minMilesExclusive: Double?
    public var maxMiles: Double?
    /// Broader travel searches rank the admitted pool together; exact miles
    /// still break ties.
    public var distanceBands: Bool
    public var bars: [Bar]
    public var excludeIds: [String]?
    public var maxResultsOverride: Int?
    public var now: Date?
    /// Live-surface clock for the LATE-NIGHT bias. Omitted on quiz/planning
    /// surfaces (they browse, not "right now").
    public var biasNow: Date?
    /// Learned taste from numeric scores. Omitted = no rating history, so
    /// c = 0 and the quiz prior alone orders the page — the cold-start case.
    public var taste: LearnedTaste

    public init(
        profile: VibeProfile,
        coords: Coords?,
        preferredNeighborhoods: [Neighborhood],
        minMilesExclusive: Double? = nil,
        maxMiles: Double?,
        distanceBands: Bool = true,
        bars: [Bar],
        excludeIds: [String]? = nil,
        maxResults: Int? = nil,
        now: Date? = nil,
        biasNow: Date? = nil,
        taste: LearnedTaste = emptyTaste
    ) {
        self.profile = profile
        self.coords = coords
        self.preferredNeighborhoods = preferredNeighborhoods
        self.minMilesExclusive = minMilesExclusive
        self.maxMiles = maxMiles
        self.distanceBands = distanceBands
        self.bars = bars
        self.excludeIds = excludeIds
        self.maxResultsOverride = maxResults
        self.now = now
        self.biasNow = biasNow
        self.taste = taste
    }
}

/// 10pm-3:59am in NEW YORK — when the night bias applies. `getHours()`
/// answers in the device's zone; this is an NYC-only matcher.
public func isLateNight(_ now: Date) -> Bool {
    let h = nycHour(now)
    if h.isNaN { return false } // broken clock: no bias rather than a wrong one
    return h >= Double(lateNightStartHour) || h < Double(lateNightEndHour)
}

/// Additive late-night nudge at tie-breaker scale: genuine nightlife
/// (club/dance) up, restaurant-bars down — unless the venue is BOTH.
public func lateNightAdjustment(_ bar: Bar) -> Double {
    let isClub = bar.tags.contains(.club) || bar.tags.contains(.dance)
    if isClub { return lateClubBoost }
    if bar.tags.contains(.restaurantBar) { return -lateRestaurantPenalty }
    return 0
}

/// Explicit-intent weighting — APPROVED product assumption (operator,
/// 2026-08-19). Do not re-tune silently; a change here is a product decision.
private let explicitVibeWeight: Double = 0.8
private let explicitTasteWeight: Double = 1 - explicitVibeWeight

/// A bar's WITHIN-BAND ordering value — the cascade's step 3:
///   c*learnedTaste + (1 - c)*quizPrior  (+ the late-night nudge)
/// Distance is deliberately ABSENT: it selects the band and breaks ties, it
/// is not a ranking term. Used when NO vibe tweak is active.
public func rankScore(_ bar: Bar, quizTags: [VibeTag], taste: LearnedTaste, late: Bool = false) -> Double {
    let c = taste.confidence
    return c * learnedTasteScore(bar, taste) + (1 - c) * jaccard(quizTags, bar.tags)
        + (late ? lateNightAdjustment(bar) : 0)
}

/// `rankScore`'s counterpart for an ACTIVE vibe tweak:
///   0.8*vibeMatch + 0.2*learnedTaste  (+ the same late-night nudge)
/// No confidence term — that is the whole point.
public func explicitVibeScore(_ bar: Bar, vibeTags: [VibeTag], taste: LearnedTaste, late: Bool = false) -> Double {
    explicitVibeWeight * jaccard(vibeTags, bar.tags) + explicitTasteWeight * learnedTasteScore(bar, taste)
        + (late ? lateNightAdjustment(bar) : 0)
}

public func matches(_ args: MatchesArgs) -> [Bar] {
    let nowForFreshness = args.now ?? Date()
    let exclude = Set(args.excludeIds ?? [])

    var pool = args.bars
        .filter { !exclude.contains($0.id) }
        // Never SUGGEST a dead bar: Places refresh found permanently-closed
        // catalog bars. Badge-only handling isn't enough — hard-filter them.
        .filter { $0.businessStatus != .closedPermanently }
        .filter { daysAgo($0.lastVerified, now: nowForFreshness) <= Double(lastVerifiedHardFilterDays) }

    if !args.preferredNeighborhoods.isEmpty {
        let allowed = Set(args.preferredNeighborhoods)
        pool = pool.filter { allowed.contains($0.neighborhood) }
    }

    // The explicit-intent path. Active only for an APPLIED tweak carrying at
    // least one tag. Resolved BEFORE the distance filter because an applied
    // pick changes what that filter admits.
    let selected = selectedVibes(args.profile)

    // ELIGIBILITY, not ordering (V8-R-NXT-009 / D-C-41). Runs BEFORE the
    // distance filter, before banding, before the cap. There is deliberately
    // no second pass that lets a rejected bar back in.
    if let selected {
        pool = pool.filter { isVibeEligible(selected, $0.tags) }
    }

    // Explicit vibes never widen a caller's geographic bounds.
    if let coords = args.coords, args.minMilesExclusive != nil || args.maxMiles != nil {
        pool = pool.filter { bar in
            let miles = haversineMiles(coords, bar.coords)
            return (args.minMilesExclusive == nil || miles > args.minMilesExclusive!)
                && (args.maxMiles == nil || miles <= args.maxMiles!)
        }
    }

    // JS `slice(0, cap)` with a negative cap is an empty list; Swift `prefix`
    // traps. Clamp once so the two agree (Codex review 2026-09-25).
    let cap = max(0, args.maxResultsOverride ?? maxResults)

    // ---- V8 P1 cascade -------------------------------------------------
    // Quiz tags are a COLD-START PRIOR, never an admission gate.
    let late = args.biasNow.map(isLateNight) ?? false

    let rankOf: (Bar) -> Double = selected != nil
        ? { bar in explicitVibeScore(bar, vibeTags: selected!, taste: args.taste, late: late) }
        : { bar in rankScore(bar, quizTags: args.profile.tags, taste: args.taste, late: late) }

    // Step 2 — fill from the CLOSEST band first, expanding only when the
    // closer band cannot fill the page. With no coords there is nothing to
    // band on, so the whole pool is one band (the pre-GPS behavior).
    let milesOf: (Bar) -> Double = args.coords != nil
        ? { bar in haversineMiles(args.coords!, bar.coords) }
        : { _ in 0 }

    var bands: [[Bar]]
    if let coords = args.coords, args.distanceBands {
        bands = [[], [], []]
        for bar in pool {
            let mi = haversineMiles(coords, bar.coords)
            let bandIndex = mi <= radiusWalk ? 0 : (mi <= radiusCab ? 1 : 2)
            bands[bandIndex].append(bar)
        }
    } else {
        bands = [pool]
    }

    // Steps 3 + 4 — within a band, learned taste orders; EXACT MILES are only
    // the final tie-breaker, never a ranking term of their own.
    var ranked: [(bar: Bar, score: Double, miles: Double)] = []
    for band in bands {
        if ranked.count >= cap { break }
        let scored = band.map { bar in (bar: bar, score: rankOf(bar), miles: milesOf(bar)) }
            .sorted { lhs, rhs in
                if lhs.score != rhs.score { return lhs.score > rhs.score }
                return lhs.miles < rhs.miles
            }
        ranked.append(contentsOf: scored)
    }

    return Array(ranked.prefix(cap)).map(\.bar)
}
